### PR TITLE
Fix network sharding eclipse attack

### PR DESCRIPTION
**Pull request by iulianpascalau**

### CODE REVIEW

File: sharding/networksharding/peerShardMapper.go [hunk 1]
```diff
@@ -113,15 +129,86 @@ func (psm *PeerShardMapper) byIDSearchingPidInFallbackCache(pid p2p.PeerID) (sha
...
+	}
+
+	pq, ok := objPidsQueue.(*pidQueue)
+	if !ok {
+		psm.pkPeerId.Remove(pk)
+		return
+	}
+
+	idxPid := pq.indexOf(pid)
+	if idxPid > -1 {
```
Comment (AdoAdoAdo) [inline sharding/networksharding/peerShardMapper.go]: please create a constant for -1 e.g idxNotFound and check against that  ( != idxNotFound ). 
you can also use maxint and make the type unsigned.

File: cmd/node/config/config.toml [hunk 2]
```diff
@@ -227,16 +227,29 @@
     Size = 75000
     Type = "LRU"
 
-[PublicKeyShardId]
+#PublicKeyShardId represents the main cache used to map Elrond block signing public keys to their associated
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/node/config/config.toml]: done

File: sharding/networksharding/pidQueue.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		data: make([]p2p.PeerID, 0),
+	}
+}
+
+func (pq *pidQueue) push(pid p2p.PeerID) {
+	pq.data = append(pq.data, pid)
+}
+
+func (pq *pidQueue) pop() p2p.PeerID {
+	evicted := pq.data[0]
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply sharding/networksharding/pidQueue.go]: in order to keep this unexported struct as simple as possible, the check has been introduced before calling pop (UpdatePeerIdPublicKey function, L179)

File: sharding/networksharding/peerShardMapper.go [hunk 4]
```diff
@@ -76,6 +90,7 @@ func (psm *PeerShardMapper) byIDWithNodesCoordinator(pid p2p.PeerID) (shardId ui
 
 	pkBuff, ok := pkObj.([]byte)
 	if !ok {
+		psm.peerIdPk.Remove([]byte(pid))
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply sharding/networksharding/peerShardMapper.go]: removed, added log.Warn as it is a programming error

File: sharding/networksharding/peerShardMapper.go [hunk 5]
```diff
@@ -99,6 +114,7 @@ func (psm *PeerShardMapper) byIDSearchingPkInFallbackCache(pkBuff []byte) (shard
 
 	shard, ok := shardObj.(uint32)
 	if !ok {
+		psm.fallbackPkShard.Remove(pkBuff)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply sharding/networksharding/peerShardMapper.go]: removed, added log.Warn as it is a programming error

File: sharding/networksharding/peerShardMapper.go [hunk 6]
```diff
@@ -113,15 +129,86 @@ func (psm *PeerShardMapper) byIDSearchingPidInFallbackCache(pid p2p.PeerID) (sha
 
 	shard, ok := shardObj.(uint32)
 	if !ok {
+		psm.fallbackPidShard.Remove([]byte(pid))
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply sharding/networksharding/peerShardMapper.go]: removed, added log.Warn as it is a programming error

File: sharding/networksharding/peerShardMapper.go (hunk 1 above)
Comment (iulianpascalau) [reply sharding/networksharding/peerShardMapper.go]: done, added const

**Pull request merged**